packages/lesswrong/server/analytics/hybridViews.ts

```diff
@@ -0,0 +1,314 @@
+   * following properties:
+   *  - A column window_end marking the end of the aggregation window. This is used to determine when to crossover
+   *    from materialized to live data. Any aggregation period will work (including no aggregation at all), but
+   *    as it happens all of the current views use daily aggregation.
+   *  - A filter on timestamp (or equivalent) against `after`. This is used in the live view to only include
+   *    data after the crossover time (which speeds up the query significantly).
+   *  - (Optional) A filter on timestamp < NOW() in the materialized version. This is to guard against bad data
+   *    (timestamps in the future) messing up the crossover time logic.
+   *
+   * Here is an example query:
+   * SELECT
+   *    count(*) AS view_count,
+   *    post_id,
+   *    (date_trunc('day', timestamp) + interval '1 second') AS window_start,
+   *    (date_trunc('day', timestamp) + interval '1 day') AS window_end
+   *  FROM
+   *    page_view
+   *  WHERE
+   *    timestamp > '${after.toISOString()}'
+   *    ${materialized ? 'AND timestamp < NOW()' : ''}
+   *  GROUP BY
+   *    post_id,
+   *    date_trunc('day', timestamp)
+   *
+   * When you select from the `virtualTable()` generated by this view you can think of it as selecting from this
+   * query, but with the timestamp logic removed.
+   */
+  queryGenerator: (after: Date, materialized: boolean) => string;
+  /**
+   * Array of functions that generate index queries, given the view name. You must provide at least one
+   * UNIQUE index, without this it isn't possible to refresh the view without locking the table. They
+   * should also all include "IF NOT EXISTS", not including this will cause errors to be logged (although
+   * it will still work).
+   */
+  indexQueryGenerators: ((viewName: string) => string)[];
+  /**
+   * The SQL client for the database you are creating the view in. At time of writing (2023-08-04) they are
+   * all in the analytics DB, but this would work in the main DB as well.
+   */
+  viewSqlClient?: RawSqlClient;
+}
+
+export class HybridView {
+  protected queryGenerator: (after: Date, materialized: boolean) => string;
+  protected indexQueryGenerators: ((viewName: string) => string)[];
+  protected versionHash: string;
+  protected identifier: string;
+  private viewSqlClient: RawSqlClient;
+  private matViewName: string;
+
+  constructor({
+    identifier,
+    queryGenerator,
+    indexQueryGenerators,
+    viewSqlClient,
+  }: HybridViewParams) {
+    const viewSqlClientToUse = viewSqlClient ?? getSqlClient();
+
+    if (!viewSqlClientToUse) throw new Error("Unable to connect to database");
+    this.viewSqlClient = viewSqlClientToUse;
+
+    this.queryGenerator = queryGenerator;
+    const allTimeQuery = queryGenerator(new Date(0), true);
+    const versionSignature = `${allTimeQuery}`;
+
+    this.versionHash = crypto.createHash("sha256").update(versionSignature).digest("hex").slice(0, 16);
```

Comment:
It's so if you change the query it will automatically create a new materialized view (it checks if the current view name exists, including this hash). This is useful for development, and in principle it ensure that when it tries to union the live query with the results from the materialzed view it should always be using the same version.

This is useful for development because it means you can basically treat the query for the hybrid view as a normal SQL query and it will create the materialized view in the background. I'm not 100% sure this is a good idea all in all though, because in prod creating the view can take a long time (hours) and so it's not something you want to do frivolously, I may change this behaviour when it's running in prod. In dev it takes less than a minute so it's fine to do

Reply:
Ohhh interesting! Seems cool to have for dev :) But could you add some comments in the code explaining what it's for?